### PR TITLE
program: support tracing of kernel modules

### PR DESCRIPTION
**Pull request by lmb**
program: support tracing of kernel modules

    Allow tracing functions in kernel modules via fentry, fexit, fmod_ret
    and tp_btf programs. The behaviour follows libbpf and is transparent
    to the user: if we can't find a target in vmlinux we attempt to find
    it in any loaded kernel module.

    Refactor TestProgramTypeLSM to test attaching via BTF. This removes
    LSM tests for BPF_F_SLEEPABLE, since they don't excercise library
    behaviour beyond passing ProgramSpec.Flags to the kernel.

    Updates #705

btf: add FindHandle

    Add a helper that iterates all BTF objects in the kernel and returns
    the first one for which a user supplied callback returns true.

btf: allow passing *Type to Spec.TypeByName

### CODE REVIEW

File: btf/handle.go [hunk 1]
```diff
@@ -119,3 +119,33 @@ func (it *HandleIterator) Next(handle **Handle) bool {
 func (it *HandleIterator) Err() error {
 	return it.err
 }
+
+// FindHandle returns the first handle for which predicate returns true.
+//
+// Requires CAP_SYS_ADMIN.
+//
+// Returns ErrNotFound if predicate never returns true or if there is no BTF
```
Comment (dylandreimerink) [inline btf/handle.go:127]: Nit: You return a wrapped version of `ErrNotFound`, this might give users the impression they can use `err != ErrNotFound` to check the output while they need to check with `error.Is()`. So I would reword the comment, remove it or change the return statement. Which ever suites best

**Review by dylandreimerink — CHANGES_REQUESTED** [review]

File: btf/handle.go [hunk 2]
```diff
@@ -119,3 +119,33 @@ func (it *HandleIterator) Next(handle **Handle) bool {
...
+
+	it := new(HandleIterator)
+	for it.Next(&handle) {
+		info, err := handle.Info()
+		if err != nil {
+			return nil, fmt.Errorf("info for ID %d: %w", it.ID, err)
+		}
+
+		if predicate(info) {
+			tmp := handle
```
[the remark this answers is not in the document]
Comment (lmb) [reply btf/handle.go:141]: The point is to avoid `defer handle.Close()` closing the handle we are returning. It is used in `findTargetInModule` as well.

Not sure how to solve this without adding the same comment to every callsite. Maybe I need to change HandleIterator to own the handle after all.

File: btf/handle.go (hunk 1 above)
Comment (lmb) [reply btf/handle.go:127]: Good point, I'll amend the doc.

File: btf/handle.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dylandreimerink) [reply btf/handle.go:141]: Ah, I see. That is a cool pattern. Changing ownership is difficult since you are returning the handle while the iterator goes out of scope.

Any reason for not using `runtime.SetFinalizer` like with other objects that hold file descriptors like maps and programs?

[the remark this answers is not in the document]
Comment (lmb) [reply btf/handle.go]: We already set a finalizer, but it's better to explicitly Close since the former is run from the GC. There is no guarantee how often the GC runs, so in the meantime leaked handles can use up fds. This in turn leads to fd exhaustion.

I've changed HandleIterator slighty and added a `Take` method which should make this trick more obvious: https://github.com/cilium/ebpf/pull/737/commits/2b8def66cc76a75d1a5643b632d2175b668b8e7c
